pandas/core/arrays/period.py

```diff
@@ -539,21 +779,215 @@ def _check_timedeltalike_freq_compat(self, other):
+        elif is_categorical_dtype(dtype):
+            return Categorical(self, dtype=dtype)
+        elif is_period_dtype(dtype):
+            return self.asfreq(dtype.freq)
+        else:
+            return np.asarray(self, dtype=dtype)
+
+    def _item(self):
+        if len(self) == 1:
+            # IndexOpsMixin will catch and re-raise IndexErrors
+            return Period._from_ordinal(self._ndarray_values[0], self.freq)
+        else:
+            raise ValueError('can only convert an array of size 1 to a '
+                             'Python scalar')
+
+    @property
+    def flags(self):
+        # TODO: remove
+        # We need this since reduction.SeriesBinGrouper uses values.flags
+        # Ideally, we wouldn't be passing objects down there in the first
+        # place.
+        return self._ndarray_values.flags
+
+    # ------------------------------------------------------------------------
+    # Ops
+
+    def _addsub_int_array(self, other, op):
+        assert op in [operator.add, operator.sub]
+        # easy case for PeriodIndex
+        if op is operator.sub:
+            other = -other
+        res_values = algos.checked_add_with_arr(self.asi8, other,
+                                                arr_mask=self._isnan)
+        res_values = res_values.view('i8')
+        res_values[self._isnan] = iNaT
+        return type(self)(res_values, freq=self.freq)
+
+
+PeriodArray._add_comparison_ops()
+PeriodArray._add_datetimelike_methods()
 
 
 # -------------------------------------------------------------------
 # Constructor Helpers
 
+def period_array(data, freq=None, ordinal=None, copy=False):
+    # type: (Sequence[Optional[Period]], Optional[Tick]) -> PeriodArray
+    """
+    Construct a new PeriodArray from a sequence of Period scalars.
+
+    Parameters
+    ----------
+    data : Sequence of Period objects
+        A sequence of Period objects. These are required to all have
+        the same ``freq.`` Missing values can be indicated by ``None``
+        or ``pandas.NaT``.
+    freq : str, Tick, or Offset
+        The frequency of every element of the array. This can be specified
+        to avoid inferring the `freq` from `data`.
+    copy : bool, default False
+        Whether to ensure a copy of the data is made.
+
+    Returns
+    -------
+    PeriodArray
+
+    See Also
+    --------
+    PeriodArray
+    pandas.PeriodIndex
+
+    Examples
+    --------
+    >>> period_array([pd.Period('2017', freq='A'),
+    ...               pd.Period('2018', freq='A')])
+    <PeriodArray>
+    ['2017', '2018']
+    Length: 2, dtype: period[A-DEC]
+
+    >>> period_array([pd.Period('2017', freq='A'),
+    ...               pd.Period('2018', freq='A'),
+    ...               pd.NaT])
+    <PeriodArray>
+    ['2017', '2018', 'NaT']
+    Length: 3, dtype: period[A-DEC]
+    """
+
+    if data is None and ordinal is None:
+        raise ValueError("Must pass one of 'data' or 'ordinal'.")
+    elif data is None:
```

Comment:
what is the purpose of ordinal? its not in the doc-string

Reply:
Passing `ordinal` triggers the same behavior as `PeriodArray.__new__(ordinal=...)`.

I'll see if this is actually used anywhere.

Reply:
Based on a brief survey, it seems like it's just used in tests. The only callers are `PeriodIndex(ordinal=...)`.

```
pandas/tests/indexes/period/test_period.py
426:        idx1 = PeriodIndex(ordinal=[-1, 0, 1], freq='A')
427:        idx2 = PeriodIndex(ordinal=np.array([-1, 0, 1]), freq='A')

```

```
pandas/tests/indexes/common.py
318:                result = index_type(ordinal=index.asi8, copy=False,

```

So the question is: do we want to give users the ability to construct a `PeriodIndex` / `PeriodArray` from an array of integer ordinals + a freq? I don't think this is likely to occur, so I'd recommend:

1. removing the `ordinal` argument from `period_array`
2. deprecating `ordinal` in the `PeriodIndex` constructor, in favor of passing a `PeriodArray`.

Reply:
yep